## 5. Running BLAST from a container

 [\<\< Episode 4](https://github.com/PawseySC/bio-workshop-18/blob/master/4.hpc.md)
 | [TOC](https://github.com/PawseySC/bio-workshop-18/blob/master/README.md) |
 [Episode 6 \>\>](https://github.com/PawseySC/bio-workshop-18/blob/master/6.cellranger.md)
______


### Before we start
To begin, we'll move to the directory:

    cd bio-workshop-18/episode5_blast

There is a script, `blast-demo.sh`, that will run all the commands if you want.  Otherwise, you
can follow along below.


---
### Looking for a bioinformatics container
Imagine we want to run a bioinformatics application using a container, 
for instance [BLAST](https://blast.ncbi.nlm.nih.gov/Blast.cgi), the Basic Local Alignment Search Tool. 

As mentioned in [Episode 1](https://github.com/PawseySC/bio-workshop-18/blob/master/1.containers.md), 
a container project of very good quality for bioinformatics is [Biocontainers](https://github.com/BioContainers/containers). 
There are two main Web registries that host images from Biocontainers:
- [QUAY Hub](https://quay.io) : `quay.io/` prefix for the repository
- [Docker Hub](https://hub.docker.com) : default with `docker`

With the web browser, go to https://hub.docker.com and use the search box on the top left to look for `blast`.  
The first hit should be `biocontainers/blast`. Click on it, then click on the *Tags* button on the top to see a list of available versions. 
At the time of writing, the most recent one has the tag `v2.2.31_cv2`.

Now, let us go back to our terminal, and `pull` the image we just identified on Docker Hub:

```
docker pull biocontainers/blast:v2.2.31_cv2
```

The output will look like:

```
Using default tag: latest
latest: Pulling from biocontainers/blast
22dc81ace0ea: Pull complete
1a8b3c87dba3: Pull complete
...
Status: Downloaded newer image for biocontainers/blast:latest
```

Note how we had to specify all the ingredients for the repository: registry, author, package, tag. 
Some bioinformatics containers in Docker Hub and most in QUAY Hub have no `latest` tag in the repositories, so the tag name always needs to be specified.

Also n that, during the Pawsey workshop, you might get messages saying `Already exists`, with the pull taking a very short time. 
This would happen because the container images were pre-loaded in the cloud virtual machines, to save time during the workshop.


---
### Inspecting the container
We can run a simple command to verify the container works:

```
docker run --rm biocontainers/blast:v2.2.31_cv2 blastp -help
```

The output is:

```
USAGE
  blastp [-h] [-help] [-import_search_strategy filename]
...
 -use_sw_tback
   Compute locally optimal Smith-Waterman alignments?
```

It is always a good idea to quickly inspect a newly pulled container, and `bash` is certainly useful to this end.
Let us run the image in interactive mode:

    docker run -it --rm biocontainers/blast:v2.2.31_cv2 bash

Now from inside the container, let us find out the version of BLAST:

    blastp -version

At the time of writing, the output is:

    blastp: 2.2.31+
    Package: blast 2.2.31, build Apr 23 2016 15:49:47


Another good question is, where has the application been installed in the container?

    which blastp

Now the output is:

    /opt/conda/bin/blastp

Knowing where the package is installed can be useful in certain situations, e.g. when package libraries need to be used.  
This BLAST container has been built using Conda.  
In other Docker Hub images applications are installed in`/usr/`.  
In contrast, in QUAY Hub most of the biocontainer images have the applications installed in `/usr/local/`.  

We can now exit the container:

    exit


---
### Aligning sample data against a database
Let's copy some data over to start working with:

```
cp -p ../data_files/P04156.fasta .
cp -p ../data_files/zebrafish.1.protein.faa.gz .
gunzip zebrafish.1.protein.faa.gz
```

This is a human prion FASTA sequence along with a reference database we'll blast against.  We need to prepare the zebrafish database with `makeblastdb`:

```
docker run --rm -v $(pwd):/data/ -w /data biocontainers/blast:v2.2.31_cv2 makeblastdb -in zebrafish.1.protein.faa -dbtype prot
```

We can now run BLAST and compare our data against the database:

```
docker run --rm -v $(pwd):/data/ -w /data biocontainers/blast:v2.2.31_cv2 blastp -query P04156.fasta -db zebrafish.1.protein.faa -out results.txt
```

The final results are stored in `results.txt`:

```
less results.txt
```

which displays:

```
                                                                     Score     E
equences producing significant alignments:                          (Bits)  Value

 XP_017207509.1 protein piccolo isoform X2 [Danio rerio]             43.9    2e-04
 XP_017207511.1 mucin-16 isoform X4 [Danio rerio]                    43.9    2e-04
 XP_021323434.1 protein piccolo isoform X5 [Danio rerio]             43.5    3e-04
 XP_017207510.1 protein piccolo isoform X3 [Danio rerio]             43.5    3e-04
 XP_021323433.1 protein piccolo isoform X1 [Danio rerio]             43.5    3e-04
 XP_009291733.1 protein piccolo isoform X1 [Danio rerio]             43.5    3e-04
 NP_001268391.1 chromodomain-helicase-DNA-binding protein 2 [Dan...  35.8    0.072
...
```


---
### Conclusion ###
There are a lot of applications (not just bioinformatics) already wrapped up in container images.  Here's a small list of some of the repositories we use at Pawsey:

* [DockerHub](hub.docker.com)
* [Bioboxes](bioboxes.org)
* [Biocontainers](biocontainers.pro)
* [Nvidia GPU Cloud (NGC)](ngc.nvidia.com)*
* [QUAY Hub](quay.io)*

The last two require you to create an account and login to access containers.


---
### Best practices ###

- Don't re-invent the wheel, it's worth looking to see who's done what you want already;
- When looking for container images for bioinformatics packages on the Web, the best places to search are [QUAY Hub](https://quay.io) and [Docker Hub](https://hub.docker.com);
- An image is copied locally from a Web repo using `docker pull <image path>`.
- A pulled image can be executed straight away with a command of the type:
  
        docker run <docker flags> <container ID> <original command>


______
 [\<\< Episode 4](https://github.com/PawseySC/bio-workshop-18/blob/master/4.hpc.md)
 | [TOC](https://github.com/PawseySC/bio-workshop-18/blob/master/README.md) |
 [Episode 6 \>\>](https://github.com/PawseySC/bio-workshop-18/blob/master/6.cellranger.md)
